### PR TITLE
libsql wal current segment

### PR DESCRIPTION
**Pull request by MarinPostma**
add current segment and transaction to the wal.

### CODE REVIEW

File: libsql-wal/src/segment/current.rs [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+            if let Some(size_after) = size_after {
+                if tx.not_empty() {
+                    let last_frame_no = tx.next_frame_no - 1;
+                    let mut header = { *self.header.lock() };
+                    header.last_commited_frame_no = last_frame_no.into();
+                    header.db_size = size_after.into();
+                    header.recompute_checksum();
+
+                    self.file.write_all_at(header.as_bytes(), 0)?;
+                    // self.file.sync_data().unwrap();
```
Comment (LucioFranco) [inline libsql-wal/src/segment/current.rs:143]: whats going on with this fsync?

**Review by LucioFranco — APPROVED** [review]
Left some feedback, can be done in follow up, nice work 👍

File: libsql-wal/src/segment/current.rs [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    /// Read lock count on this segment. Each begin_read increments the count of readers on the current
+    /// lock
+    read_locks: Arc<AtomicU64>,
+    sealed: AtomicBool,
+    tail: Arc<SegmentList<F>>,
+}
+
+impl<F> CurrentSegment<F> {
+    /// Create a new segment from the given path and metadata. The file pointed to by path must not
+    /// exist.
```
Comment (avinassh) [inline libsql-wal/src/segment/current.rs:37]: how about `assert`ing this assumption in `create`

**Review by avinassh — APPROVED** [review]

**Review by avinassh — APPROVED** [review]

File: libsql-wal/src/segment/current.rs [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+                    header.recompute_checksum();
+
+                    self.file.write_all_at(header.as_bytes(), 0)?;
+                    // self.file.sync_data().unwrap();
+                    tx.merge_savepoints(&mut self.index.index.write());
+                    // set the header last, so that a transaction does not witness a write before
+                    // it's actually committed.
+                    *self.header.lock() = header;
+
+                    tx.is_commited = true;
```
[the remark this answers is not in the document]
Comment (MarinPostma) [reply libsql-wal/src/segment/current.rs:149]: This is not how SQLite works, SQLite commits the transaction in the call to insert. ´end-write ´ just releases the lock at the end.

[the remark this answers is not in the document]
Comment (avinassh) [reply libsql-wal/src/segment/current.rs:149]: TIL! That also explains some other questions I had related to the locks and write transactions.

**Comment by MarinPostma** [comment]
will nits comments as followups

**Pull request merged**